### PR TITLE
kernel: device: add new driver define interface for abstract platform complexity from device drivers

### PR DESCRIPTION
**Pull request by najumon1980**
add new driver define interface for abstract platform complexity from device drivers.

**Issue/problem Statement.**
======================================

Zephyr RTOS introduced many features, spatially on driver framework which make it much more capable from other counter part RTOS in case of scalability, configurability, modularity and re-usability. Generic APIs for IO drivers, device tree, power management, user mode, PCI bus driver etc. which make it capable for supporting in numerous use case from very tiny low foot print MCU to Application processor based platforms with complex sub subsystems. But as OS become more capable it become more complex as well. Following are some of the limitations of existing Zephyr driver interface:

- Developers have to understand and implement various platform and OS specific interface code such PCI bus, DTS, interrupt configuration/management etc. in each drivers they newly write or even modify an existing driver for porting into their platform. 
- The porting will be more complex when device is sitting on some of the bus such as PCI or ACPI
- These OS/Bus/platform interface might change with platforms as well as newer Zephyr versions and hence we need to maintain/modify every driver source code for migrate to new platform/OS version.
- Less re-usable: These bus/platform adoption code almost have similar logic for resource enumeration, interrupt configuration etc. and so other drivers can easily re-use them once developed for one IO device. But with current model developers have to add these code into every drivers they develop/port.

If we can enhance existing driver interface in such a way that most of these platform complexities can be abstracted behind interface, then driver developers can focus only on device/ip specific implementation code for their driver. 

> **Note**
Please note that the PR is only a draft version and currently given for analyze purpose. 

**How to enhance interface for IO drivers - Draft Proposal**
======================================
Following section brief on the new proposal for enhance device driver interface. The proposal is only able to capture base requirements, but might need continue improvement by review and adopt suggestion from community and domain experts. 

**Analyze on Architecture Changes on Existing driver Interface**
------------------------------------------------------------------
As shown in the below block diagram, the proposed model have abstract interface for platform specific implementation and other bus specific complexity and provide a unified interface to driver developers. It also propose interface for IO subsystem such as i2c driver, spi driver etc. The sensor bus interface can further abstract sensor drivers from underlying bus interface such as i2c, spi etc.

image

**Platform Abstract Interface**
-------------------------------

In the proposed model, Zephyr provide a new driver define interface for drivers called "PLATFORM_DEV_DT_INST_DEFINE".

> **Note**
In order to make backsword compatible with existing drivers, the new interface by default will be disabled and can be enabled only if developers have to use this new interface for their drivers. Also the new interface will co-exist with existing DEVICE_DT_DEFINE interface. 

Following code snippet depicts how this new interface can be used for driver development .

```diff
static void my_dev_isr(const void *arg)
{
	const struct device *dev = arg;
	
+	platform_write_reg_32(dev, MY_REG_OFFSET_2, MY_INTR_CFG_1);

	...

}

static int my_dev_config(const struct device *dev,
			struct my_config *config_param)
{
	const struct my_device_config * config = dev->config;
	struct my_device_data * data = dev->data;
	
+	platform_write_reg_32(dev, MY_REG_OFFSET_1, config_param->my_cfg1);

	...

}

static int my_dev_initialize(const struct device *dev)
{
	const struct my_device_config * config = dev->config;
	struct my_config *default_cfg = config->my_default_config;
	
+	platform_write_reg_32(dev, MY_REG_OFFSET_1, default_cfg->my_cfg1);

	...
	
}

struct my_device_api my_dev_api = {
	.my_config = my_dev_config,
	...
};

#define MY_DEVICE_INIT(n) \
	static const struct my_device_config my_dev_config_##n = {           \
		.my_default_config = DT_INST_PROP(n, my_some_config),                  \
	};                                                                    \
	static struct my_device_data my_dev_data_##n;                    \
+	PLATFORM_DEV_DT_INST_DEFINE(n, BOOT_POST_KERNEL, NULL, my_dev_initialize,\
+		NULL, &my_dev_data_##n, &my_dev_config_##n, &my_dev_api, my_dev_isr);   \

DT_INST_FOREACH_STATUS_OKAY(MY_DEVICE_INIT)
```
Here the new PLATFORM_DEV_DT_INST_DEFINE interface is almost similar to existing DEVICE_DT_INST_DEFINE interface, except following changes:
- Boot priority limited to pre-kernel, post-kernel or application level which are based on the specific driver implementation ( example: if driver have to use some kernel APIs during init then it should be post kernel etc.) and use case (example: UART driver might required to be pre-kernel for enable early logging). These flags expected to be remain same for a specific driver and will not change from platform to platform. Device boot priority will be also depend on bus/parent driver boot priority as well apart  from these boot flag. All other inter driver dependency and boot priority will be take care by platform abstraction layer based on your device position in the device tree such as order of device node, parent node up on which the child node located etc. The framework will ensure the parent node init will be invoked before the child node init call. Similarly device init will be called in the ascending order of nodes in the device tree. This will help a platform integrator to configure the driver boot order dependency based on his platform boot order requirement, bus upon which the device node placed etc. Following diagram depict the parent child hierarchy and the device boot order. Please note that below diagram assume all driver have same boot priority say pre-kernel and if it it have different boot flag then all pre-kernel device will get instantiated first followed by post-kernel and finally application level.
 
image

- defaut_cfg (optional) which can be used for sub system interface such as i2c or spi to keep the default device configuration such as clock frequency, baud rate etc.  [Please refer I2C PR](https://github.com/zephyrproject-rtos/zephyr/pull/57765/commits/c01c80d867971bb9735bb6a51e29bc85367def31) and  [UART PR](https://github.com/zephyrproject-rtos/zephyr/pull/57765/commits/c01c80d867971bb9735bb6a51e29bc85367def31)  for more detail on changes required for i2c  and UART drivers with new interface. 

- isr (optional) can be provided to the platform abstraction layer and it will take care of how to register ISR with platform or bus driver (such as PCI etc.) based on DTS, PCI information etc.

Additionally platform abstract layer will provide device register access API where driver developers need to provide only register offset. Based on SOC/platform HW, the register will be re-directed as memory mapped access or IO access. Platform abstract layer will retrieve MMIO address from either DTS or from bus such as PCI depend on the node position in the device tree and also map the MMIO address into virtual address if MMU enabled in the platform.

**Bus/Parent device**

For drivers such as SPI, I2C etc. where child device/node (sensors) can use platform abstract macro such as DEVICE_PROBE_CHILD_NODE_POST_KERNEL() to enumerate its child devices. This should be part of SPI/i2C common layer(i2c.h or spi.h) and should be abstract from device drivers.

**Platform Abstract Layer Configuration Details**

There are three type of device configuration supported:

- Generic device configuration where default configuration will be used such as IRQ registration using IRQ_CONNECT(), retrieve  MMIO address using DT_REG_ADDR etc. This can be used for MCUs where there is no constrain in ISR name/location and no MMU support etc. [Please refer PLATFORM_GENERIC_DEV_CONFIG macro in platform.h ](https://github.com/zephyrproject-rtos/zephyr/pull/57765/commits/319b3c53422d019683cc5eaddecf94f78a75a40d) for more details on generic configuration implementation.

- Platform specific configuration (PLATFORM_SPECIFIC_DEV_INIT) can be use if generic device configuration cannot be used in such case where device ISR/MMIO interface will be specific to platform. In this case there is a default week function implementation of platform_specific_dev_config() which can be used if your constrain is only dynamic IRQ allocations and MMU support . If platform require more specific implementation then SOC/platform vendor can implement their own impletion of platform_specific_dev_config() function which override the default implementation.  [Please refer platform.c ](https://github.com/zephyrproject-rtos/zephyr/pull/57765/commits/319b3c53422d019683cc5eaddecf94f78a75a40d) for more details on Platform specific week implementation.

- Bus specific configuration is for device sitting on PCI or ACPI node. In this case, device resource allocation will be taken care by underlying bus driver.  [Please refer pcie.h ](https://github.com/zephyrproject-rtos/zephyr/pull/57765/commits/ecf2b9dfe1d309c85d228cc15b6666c4624c0b8b) for more details on PCI specific implementation.

**IO Abstract Interface**
------------------------

In the proposed model, Zephyr provide a set of new driver define interface for IO drivers such as  I2C_PLATFORM_DEV_DT_DEFINE(), UART_PLATFORM_DEV_DT_DEFINE() etc.  These interface will internally use PLATFORM_DEV_DT_INST_DEFINE() to abstract underlying platform and bus interface complexity and also will fetch additional device configuration from DTS such as clock speed, baud rate etc. and provide those information through simplified interface such as I2C_SPEED_GET(), UART_DEFAULT_CONF_GET() etc.

It also do device enumeration if the DTS node have additional child node such as sensor devices and enable them for ready to use.

Following code snippet show how the new I2C_PLATFORM_DEV_DT_DEFINE() i2c interface can be used to write/port i2c device driver. The interface is almost similar to PLATFORM_DEV_DT_INST_DEFINE(). Driver can use default i2c bus configuration such as clock frequency using I2C_SPEED_GET() macro. Also you could notice a lot of code is not required in drivers since it is taken care by the interface layer.

```diff

...

+     clk_freq = I2C_SPEED_GET(dev);
+	platform_write_reg_32(dev, DW_CFG_REG, clk_freq);

...

- #if DT_ANY_INST_ON_BUS_STATUS_OKAY(pcie)
-	if (rom->pcie) {
-		struct pcie_bar mbar;
-
-		if (rom->pcie->bdf == PCIE_BDF_NONE) {
-			return -EINVAL;
-		}
-
-		pcie_probe_mbar(rom->pcie->bdf, 0, &mbar);
-		pcie_set_cmd(rom->pcie->bdf, PCIE_CONF_CMDSTAT_MEM, true);
-
-		device_map(DEVICE_MMIO_RAM_PTR(dev), mbar.phys_addr,
-			   mbar.size, K_MEM_CACHE_NONE);
-	} else
- #endif
-	{
-		DEVICE_MMIO_MAP(dev, K_MEM_CACHE_NONE);
-	}

...

- #define I2C_DW_INIT_PCIE0(n)
- #define I2C_DW_INIT_PCIE1(n) DEVICE_PCIE_INST_INIT(n, pcie),
- #define I2C_DW_INIT_PCIE(n) \
-	_CONCAT(I2C_DW_INIT_PCIE, DT_INST_ON_BUS(n, pcie))(n)
-
- #define I2C_DEFINE_PCIE0(n)
- #define I2C_DEFINE_PCIE1(n) DEVICE_PCIE_INST_DECLARE(n)
- #define I2C_PCIE_DEFINE(n) \
-	_CONCAT(I2C_DEFINE_PCIE, DT_INST_ON_BUS(n, pcie))(n)
-
- #define I2C_DW_IRQ_FLAGS_SENSE0(n) 0
- #define I2C_DW_IRQ_FLAGS_SENSE1(n) DT_INST_IRQ(n, sense)
- #define I2C_DW_IRQ_FLAGS(n) \
-	_CONCAT(I2C_DW_IRQ_FLAGS_SENSE, DT_INST_IRQ_HAS_CELL(n, sense))(n)
-
- /* not PCI(e) */
- #define I2C_DW_IRQ_CONFIG_PCIE0(n)                                            \
-	static void i2c_config_##n(const struct device *port)                 \
-	{                                                                     \
-		ARG_UNUSED(port);                                             \
-		IRQ_CONNECT(DT_INST_IRQN(n), DT_INST_IRQ(n, priority),	      \
-			    i2c_dw_isr, DEVICE_DT_INST_GET(n),                \
-			    I2C_DW_IRQ_FLAGS(n));                             \
-		irq_enable(DT_INST_IRQN(n));                                  \
-	}
-
- /* PCI(e) with auto IRQ detection */
- #define I2C_DW_IRQ_CONFIG_PCIE1(n)                                            \
-	static void i2c_config_##n(const struct device *port)                 \
-	{                                                                     \
-		BUILD_ASSERT(DT_INST_IRQN(n) == PCIE_IRQ_DETECT,              \
-			     "Only runtime IRQ configuration is supported");  \
-		BUILD_ASSERT(IS_ENABLED(CONFIG_DYNAMIC_INTERRUPTS),           \
-			     "DW I2C PCI needs CONFIG_DYNAMIC_INTERRUPTS");   \
-		const struct i2c_dw_rom_config * const dev_cfg = port->config;\
-		unsigned int irq = pcie_alloc_irq(dev_cfg->pcie->bdf);        \
-		if (irq == PCIE_CONF_INTR_IRQ_NONE) {                         \
-			return;                                               \
-		}                                                             \
-		pcie_connect_dynamic_irq(dev_cfg->pcie->bdf, irq,	      \
-				     DT_INST_IRQ(n, priority),		      \
-				    (void (*)(const void *))i2c_dw_isr,       \
-				    DEVICE_DT_INST_GET(n),                    \
-				    I2C_DW_IRQ_FLAGS(n));                     \
-		pcie_irq_enable(dev_cfg->pcie->bdf, irq);                     \
-	}
-
- #define I2C_DW_IRQ_CONFIG(n) \
-	_CONCAT(I2C_DW_IRQ_CONFIG_PCIE, DT_INST_ON_BUS(n, pcie))(n)
-
- #define I2C_CONFIG_REG_INIT_PCIE0(n) DEVICE_MMIO_ROM_INIT(DT_DRV_INST(n)),
- #define I2C_CONFIG_REG_INIT_PCIE1(n)
- #define I2C_CONFIG_REG_INIT(n) \
-	_CONCAT(I2C_CONFIG_REG_INIT_PCIE, DT_INST_ON_BUS(n, pcie))(n)
-
 #define I2C_DEVICE_INIT_DW(n)                                                 \
	PINCTRL_DW_DEFINE(n);                                                 \
-	I2C_PCIE_DEFINE(n);                                                   \
-	static void i2c_config_##n(const struct device *port);                \
-	static const struct i2c_dw_rom_config i2c_config_dw_##n = {           \
-		I2C_CONFIG_REG_INIT(n)                                        \
-		.config_func = i2c_config_##n,                                \
-		.bitrate = DT_INST_PROP(n, clock_frequency),                  \
		PINCTRL_DW_CONFIG(n)                                          \
-		I2C_DW_INIT_PCIE(n)                                           \
-	};                                                                    \
	static struct i2c_dw_dev_config i2c_##n##_runtime;                    \
-	I2C_DEVICE_DT_INST_DEFINE(n, i2c_dw_initialize, NULL,                 \
-			      &i2c_##n##_runtime, &i2c_config_dw_##n,         \
			      POST_KERNEL, CONFIG_I2C_INIT_PRIORITY,          \
-			      &funcs);                                        \
-	I2C_DW_IRQ_CONFIG(n)
-
...

+	I2C_PLATFORM_DEV_DT_DEFINE(n, BOOT_POST_KERNEL, \
                i2c_dw_initialize,\
		NULL, &i2c_##n##_runtime, &i2c_config_dw_##n, &funcs,\
+             i2c_dw_isr);   \

...

```
**Sensor Bus Abstract Interface**
-----------------------------
TODO

### CODE REVIEW

**Comment by najumon1980** [comment]
hi maintainers, could you please move this PR to DNM

File: cmake/linker_script/common/common-rom.cmake [hunk 1]
```diff
@@ -7,6 +7,7 @@ zephyr_linker_section_obj_level(SECTION init LEVEL PRE_KERNEL_2)
 zephyr_linker_section_obj_level(SECTION init LEVEL POST_KERNEL)
 zephyr_linker_section_obj_level(SECTION init LEVEL APPLICATION)
 zephyr_linker_section_obj_level(SECTION init LEVEL SMP)
+zephyr_linker_section_obj_level(SECTION init LEVEL CUSTOM_BOOT_LEVEL)
```
Comment (gmarull) [inline cmake/linker_script/common/common-rom.cmake:10]: Can you explain why we need yet another init level? How is this related with this proposal?

File: dts/bindings/base/base.yaml [hunk 2]
```diff
@@ -19,6 +19,18 @@ properties:
...
+    type: string
+    description: Device hardware id
+
+  unique-id:
+    type: string
+    description: Device instance number
+
+  class-code:
+    type: string
+    description: Device class id
```
Comment (gmarull) [inline dts/bindings/base/base.yaml:32]: this is not hardware description, and so it should not be in devicetree.

File: dts/bindings/base/base.yaml [hunk 3]
```diff
@@ -19,6 +19,18 @@ properties:
     required: true
     description: compatible strings
 
+  hardware-id:
+    type: string
+    description: Device hardware id
+
+  unique-id:
+    type: string
+    description: Device instance number
```
Comment (gmarull) [inline dts/bindings/base/base.yaml:28]: same here

File: kernel/platform.c [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+#include <string.h>
+#include <zephyr/device.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/sys/kobject.h>
+#include <zephyr/syscall_handler.h>
+#if defined(CONFIG_PLATFORM_ABST_SUPPORT)
+#include <zephyr/sys/slist.h>
+#include <zephyr/irq.h>
+#include <zephyr/device.h>
```
Comment (gmarull) [inline kernel/platform.c:15]: duplicate device includes

File: kernel/platform.c [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+#include <zephyr/platform.h>
+#endif
+
+#define LOG_LEVEL CONFIG_PLATFORM_ABST_SUPPORT_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(platform_abst);
+
+#if defined(CONFIG_PLATFORM_ABST_SUPPORT)
+__attribute__((weak)) int platform_specific_dev_config(const struct device *dev,
```
Comment (gmarull) [inline kernel/platform.c:24]: why weaks here?

File: kernel/platform.c [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+		}
+		if (status > UINT8_MAX) {
+			status = UINT8_MAX;
+		}
+		dev->state->init_res = status;
+	}
+
+	dev->state->initialized = true;
+
+	return 0;
```
Comment (gmarull) [inline kernel/platform.c:74]: this seems to duplicate logic already present for _regular_ devices

File: drivers/pcie/host/pcie.c [hunk 7]
```diff
@@ -22,6 +23,9 @@ LOG_MODULE_REGISTER(pcie, LOG_LEVEL_ERR);
 #include <zephyr/drivers/pcie/controller.h>
 #endif
 
+#define PCIE_LOG_DBG(...) printk(__VA_ARGS__)
```
Comment (gmarull) [inline drivers/pcie/host/pcie.c:26]: ?

File: drivers/serial/uart_ns16550.c [hunk 8]
```diff
@@ -330,79 +316,38 @@ struct uart_ns16550_dev_data {
...
 {
-#if defined(CONFIG_UART_NS16550_ACCESS_IOPORT) || defined(CONFIG_UART_NS16550_SIMULT_ACCESS)
-	if (cfg->io_map) {
-		if (IS_ENABLED(CONFIG_UART_NS16550_ACCESS_WORD_ONLY)) {
-			sys_out32(val, port);
-		} else {
-			sys_out8(val, port);
-		}
+	if (IS_ENABLED(CONFIG_UART_NS16550_ACCESS_WORD_ONLY)) {
+		platform_write_reg_32(dev, port, (uint32_t)val);
```
Comment (gmarull) [inline drivers/serial/uart_ns16550.c:323]: have you considered looking at eg Linux regmap?

**Review by gmarull — CHANGES_REQUESTED** [review]
I think this PR should be reduced to something minimal, and justify the changes better. For example, a typical thing we do is to compare with Linux.

**Review by tbursztyka — CHANGES_REQUESTED** [review]
I believe the issue is wrongly stated: truly, the bottom of the problem is how convoluted i2c_dw and ns16550 (perhaps a tiny few others) are coded due to various context they have to run competitively, on pcie, on iomap, etc... So far, these are kind of an exception among drivers, thus the given solution is not entitled to be generalized into the device driver model. For instance, being io-mapped or not is really specific and not arch-agnostic. Not to mention that it bloats the struct device way too much then. (ROM/RAM usage is something important to consider in Zephyr)

Of course, there is also the boot ordering issue. That's a well known one, and so far no proper solution have been found. And the proposed PR here is not the right one either. I am not going to dissect the proposed solution, just see #22545 and the various PR attempts that followed (main issue being that DTS is purely hw-centric while the init order does not only affect hw but also sw services, and inter-dependency between the two)

Then there are all the technical non-Zephyr ways. Devices are already uniquely identified through DTS: no need to add another one via an hw id that one would have to create manually (again that does not solve the sw services and so on). Runtime device dependency tree already exists (see the attribute device_handle_t *handles in struct device and the related API around), no need to artificially recreate one. Also, let's not store information that is meant to be used at build time (irq num/prio/flag for instance among others)

Finally, I did not go into a detailed review since right from the start there are major mistakes that make this PR non build-able at this stage: there is no .init nor .isr attribute in a struct init_entry for instance, structs are not properly aligned, so many typos etc...

**Comment by tbursztyka** [comment]
> a typical thing we do is to compare with Linux.

Not only linux, but BSD as well or any other open source OS around. That being said it's good to look around with a grain of salt as there are reasons why things are done a certain way in every OS and sometime (quite often actually) they just do not fit Zephyr use cases. Like the init order for instance, we really want something build-base to save boot time and most importantly: memory.

File: cmake/linker_script/common/common-rom.cmake (hunk 1 above)
Comment (najumon1980) [reply cmake/linker_script/common/common-rom.cmake]: updated PR to get child device information via DT_FOREACH_CHILD_STATUS_OKAY_SEP and DEVICE_INIT_DT_GET macros. This will avoid walking through the section for getting right device object (using Hardware id etc.). So this is no more required and is removed.

File: dts/bindings/base/base.yaml (hunk 2 above)
Comment (najumon1980) [reply dts/bindings/base/base.yaml]: No more required and is removed with updated code

File: kernel/platform.c (hunk 5 above)
Comment (najumon1980) [reply kernel/platform.c:17]: This is required in case SOC/platform vendors need to implement their own config method. The function prototype might need further update as this PR is very initial draft version and need to align the architecture scope before going to detailed implementation.

File: drivers/pcie/host/pcie.c (hunk 7 above)
Comment (najumon1980) [reply drivers/pcie/host/pcie.c]: removed
